_posts/2022-06-10-slick-intro.md

```diff
@@ -0,0 +1,377 @@
+      extends API
+      with JsonImplicits
+      with HStoreImplicits
+      with ArrayImplicits
+      with DateTimeImplicits {
+    implicit val strListTypeMapper = new SimpleArrayJdbcType[String]("text").to(_.toList)
+    implicit val playJsonArrayTypeMapper =
+      new AdvancedArrayJdbcType[JsValue](
+        pgjson,
+        (s) => utils.SimpleArrayUtils.fromString[JsValue](Json.parse(_))(s).orNull,
+        (v) => utils.SimpleArrayUtils.mkString[JsValue](_.toString())(v)
+      ).to(_.toList)
+  }
+}
+object CustomPostgresProfile extends CustomPostgresProfile
+```
+Now, we can use `CustomPostgresProfile` instead of the `PostgresProfile` to make use of these features.
+
+### 6.1. Querying from an Array Column
+
+Let's see how we can use filter query on an postgres array column. We have created a _MovieLocationsTable_ which has a _movieLocations_ array field. If we want to filter movies which was shot on any of the input locations, we can do that as:
+```scala
+val locations: List[String] = List("location1", "location2")
+val query = SpecialTables.movieLocationsTable.filter(_.locations @& locations.bind)
+```
+The operator `@&` will return true if there is an overlap between the input list and the database column. There are many other such operators on array field, but we are not going to go through all of it here.
+
+### 6.2. Querying from an HStore Column
+
+Postgres has an extension called as `hstore`. Once it is installed, we can use the datatype _hstore_ for the columns. It is a type which stores the data as key and value pair, equivalent to an _Map_ type in Scala/Java. We can see how to filter a hstore column using slick-pg:
+
+```scala
+def getMoviesByDistributor(distributor: String): Future[Seq[MovieProperties]] = {
+  val condition = Map("distributor" -> distributor)
+  val query = SpecialTables.moviePropertiesTable.filter(_.properties @> condition.bind)
+  Connection.db.run(query.result)
+}
+```
+The operator `@>` filters the hstore column for the key _distributor_ and the input value. Similar to array, there are many other operators in hstore as well.
+
+### 6.3. Querying from a JSON Column
+Postgres also supports JSON datatype by default. Slick-Pg also has support for querying the json columns. Let's look at it with an example:
+
+```scala
+def getActorsBornOn(year: String): Future[Seq[ActorDetails]] = {
+  Connection.db.run(
+    SpecialTables.actorDetailsTable.filter(_.personal.+>>("birthYear") === year.bind).result
+  )
+}
+```
+In the above code, we are filtering the _personal_ column which is a JSON type in _ActorDetails_ table. The method `+>>` will get the json key _birthYear_ and compares it with the value provided.
+
+## 7. Code Generation
+So far, we have written the Slick tables manually. If we are following a database first approach, slick provides a way to generate the mapping tables easily. For now, we are ignoring the slick-pg types and trying to generate the slick mappings for basic datatypes. 
```

Comment:
We need to add a few more bits to teach the user how to generate the table files, where to look for them and what their structure is, in order for them to understand what's being generated and how to use them.